apps/designer/app/routes/designer/$id.tsx

```diff
@@ -1,30 +1,82 @@
 import { useLoaderData } from "@remix-run/react";
-import { LoaderFunction } from "@remix-run/node";
-import type { Project } from "@webstudio-is/react-sdk";
+import path from "path";
+import {
+  ActionFunction,
+  LoaderFunction,
+  unstable_createFileUploadHandler,
+  unstable_parseMultipartFormData,
+} from "@remix-run/node";
+import type { Project, Asset } from "@webstudio-is/react-sdk";
 import { Designer, links } from "~/designer";
 import * as db from "~/shared/db";
 import config from "~/config";
 import env from "~/env.server";
+import { z } from "zod";
 export { links };
 
 export const loader: LoaderFunction = async ({ params }) => {
   if (params.id === undefined) throw new Error("Project id undefined");
   const project = await db.project.loadById(params.id);
+  const assets = await db.assets.loadByProject(params.id);
   if (project === null) {
     return { errors: `Project "${params.id}" not found` };
   }
-  return { config, project, env };
+  return { config, assets, project, env };
 };
 
 type Data = {
   config: typeof config;
   project: Project;
+  assets: Asset[];
 };
 
 type Error = {
   errors: "string";
 };
 
+const ImageUpload = z.object({
+  name: z.string(),
+  type: z.string(),
+});
+type ImageUpload = z.infer<typeof ImageUpload>;
+
+export const action: ActionFunction = async ({ request, params }) => {
+  const uploads = path.join(__dirname, "../public");
+  const folderInPublic = process.env.FILE_UPLOAD_PATH || "uploads";
+  const directory = path.join(uploads, folderInPublic);
+
+  if (params.id === undefined) throw new Error("Project id undefined");
+  try {
+    const formData = await unstable_parseMultipartFormData(
+      request,
+      unstable_createFileUploadHandler({
+        maxPartSize: 10_000_000,
+        directory,
+        file: ({ filename }) => filename,
+      })
+    );
+
+    const imageInfo = ImageUpload.parse(formData.get("image"));
+
+    const data = {
+      type: imageInfo.type,
+      name: imageInfo.name,
+      path: `/${path.join(folderInPublic, imageInfo.name)}`,
```

Comment:
why do you wanna put path into database? we can always have a path based on the file name

Reply:
is imageInfo.name having a unique part in it?

Reply:
Yes, that is unique but that's mostly because then to show it in the frontend you don't have to check if the user is using S3 or wtv you can just have the path as src

Reply:
In the frontend you want to show the image name, not the path, right?

Reply:
yes